include/triton/backend/device_memory_tracker.h

```diff
@@ -0,0 +1,302 @@
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+#include <map>
+#include <memory>
+#include <mutex>
+#include <unordered_map>
+#include <vector>
+#include <algorithm>
+#include <functional>
+#include "triton/common/logging.h"
+#include "triton/core/tritonserver.h"
+#include "triton/backend/backend_common.h"
+
+#ifdef TRITON_ENABLE_GPU
+#include <cupti.h>
+#endif
+
+static_assert(
+    sizeof(uint64_t) >= sizeof(uintptr_t),
+    "The implementation is storing address pointer as uint64_t, "
+    "must ensure the space for pointer is <= sizeof(uint64_t).");
+
+namespace triton { namespace backend {
+
+/// DeviceMemoryTracker is a backend utility provided to track the memory
+/// allocated for a particular model and associated model instance. This utility
+/// is often used for backend that set
+/// TRITONBACKEND_BackendAttributeSetEnableMemoryTracker to true, which provides
+/// additional information to Triton for making decision on model scaling and
+/// deployment.
+///
+/// Caveat: The memory tracker is implemented with CUPTI library which currently
+/// only supports single client/subscriber. This is an known limitation and as a
+/// result, the memory tracker can cause unexpected application failure if other
+/// component of the Triton process also uses CUPTI with a different
+/// configuration, for example, the framework used by the backend may have
+/// implemented similar profiler with CUPTI. Therefore, before enabling this
+/// memory tracker utilities, you should make sure that there is no other CUPTI
+/// client in the process. This tracker is implemented with the assumption that
+/// all other possible CUPTI clients are using the same implementation so that
+/// as long as all backends are compiled with this memory tracker and they may
+/// interact with CUPTI initialized externally to the backend without issues.
```

Comment:
This sentence reads weirdly to me. How does this sound?
```suggestion
/// client in the process. This tracker is implemented with the assumption that
/// all other CUPTI clients are using the same implementation so that
/// as long as all backends are compiled with this memory tracker, they may
/// interact with an externally-initialized CUPTI without issues.
```